packages/@aws-cdk/cfnspec/build-tools/validate-cfn.ts

```diff
@@ -0,0 +1,312 @@
+/* eslint-disable no-console */
+/**
+ * CloudFormation spec validator
+ *
+ * "Why not JSON Schema?", you might ask, and it's a fair question. The answer is:
+ * because the error reporting from JSON schema is pretty bad, and I want the validation
+ * errors reported from this check to be blindingly obvious, as non-spec-experts
+ * are going to have to consume and understand them.
+ *
+ * I tried JSON Schema validation and its errors look like:
+ *
+ * ```
+ * - instance.PropertyTypes["..."].Properties.Xyz does not match allOf schema <#/definitions/Typed> with 7 error[s]:
+ * - instance.PropertyTypes["..."].Properties.Xyz requires property "PrimitiveType"
+ * - instance.PropertyTypes["..."].Properties.Xyz requires property "Type"
+ * - instance.PropertyTypes["..."].Properties.Xyz requires property "Type"
+ * - instance.PropertyTypes["..."].Properties.Xyz requires property "ItemType"
+ * - instance.PropertyTypes["..."].Properties.Xyz requires property "Type"
+ * - instance.PropertyTypes["..."].Properties.Xyz requires property "PrimitiveItemType"
+ * - instance.PropertyTypes["..."].Properties.Xyz is not exactly one from "Primitive","Complex Type","Collection of Primitives","Collection of Complex Types"
+ * ```
+ *
+ * No bueno.
```

Comment:
`~` Might be nice to instead (or also) show what this outputs, so we can see how much bueno-er this is (and why it's worth maintaining this extra ~300 line file).